pkgs/development/tools/reindeer/default.nix

```diff
@@ -0,0 +1,51 @@
+{ lib
+, rustPlatform
+, fetchFromGitHub
+, pkg-config
+, openssl
+, libiconv
+, darwin
+, stdenv
+, nix-update-script
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "reindeer";
+  version = "unstable-2023-05-18";
+
+  src = fetchFromGitHub {
+    owner = "facebookincubator";
+    repo = pname;
+    rev = "2d6d37dcddaa840b6cd99951176c972f4c3afddd";
+    sha256 = "sha256-31Kf6cVqAVFprTHEJp06E0p0y5YwdowhmPti8JBJw3g=";
+  };
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
+
+  postPatch = ''
+    ln -s ${./Cargo.lock} Cargo.lock
+  '';
+
+  doCheck = false;
+
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs =
+    [ openssl ] ++ lib.optionals stdenv.isDarwin [
+      libiconv
+      darwin.apple_sdk.frameworks.Security
+      darwin.apple_sdk.frameworks.CoreServices
+    ];
+
+  meta = with lib;
+    {
+      description = "Reindeer is a tool which takes Rust Cargo dependencies and generates Buck build rules";
+      homepage = "https://github.com/facebookincubator/reindeer";
+      license = with licenses; [ mit ];
+      maintainers = with maintainers; [ nickgerace ];
+    };
+
+  passthru.updateScript = nix-update-script { };
```

Comment:
@figsoda thanks for yourself on the other PR! I navigated to `update.nix` and other examples to try to find how to use this since [there are not docs for this yet](https://github.com/NixOS/nixpkgs/issues/216321). I _believe_ I might not need to do more here, but I want to be sure. I'd like the package to auto update the following attributes...

- `version` (update to `unstable-YYYY-MM-DD`)
- `src.rev` (update to latest commit sha on `main` for the source repo)
- `src.sha256` (update accordingly, based on `src.rev`)

...and then `Cargo.lock` needs to be regenerated. I did [something similar in my `buck2` branch here](https://github.com/nickgerace/nixpkgs/blob/48feccd207c4c58a91ebdeafbfff402210c9f092/pkgs/development/tools/build-managers/buck2/update.sh#L23-L31), but that uses the `update.sh` pattern.

Does this require any further configuration? My guess is "yes", but I'm still looking.

Reply:
for some background since there's no docs for it yet, nix-update-script is defined [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/common-updater/nix-update.nix) and uses [nix-update](https://github.com/Mic92/nix-update) under the hood. Just `nix-update-script { }` (roughly equivalent to running `nix-update reindeer`) should be sufficient here without any further configuration

Reply:
Huh that is super cool. I can't believe I don't have to do anything else here. Will have to dive deep and learn more about how it works.

These might be naive, but I have a few more questions:

- Will the @r-ryantm bot pick this up for automatic updates? Automatic updates would be ideal.
- Should running `nix-update reindeer` work locally despite the package not being added to nixpkgs yet?

Reply:
> Will the @r-ryantm bot pick this up for automatic updates?

Yes

> Should running `nix-update reindeer` work locally despite the package not being added to nixpkgs yet?

Yes, you should also be able to do `nix-shell maintainers/scripts/update.nix --argstr package reindeer`, this is how `passthru.updateScript` is normally used. Running `nix-shell maintainers/scripts/update.nix` will show you more instructions

Reply:
I see that makes sense! Looks like I need to provide it the version so that it knows what to do though...
```
nix_update.errors.VersionError: Please specify the version. We can only get the latest version from codeberg/crates.io/gitea/github/gitlab/notabug/pypi/savannah/sourcehut/rubygems projects right now
```

Reply:
this should work then

```nix
  passthru.updateScript = nix-update-script {
    extraArgs = [ "--version=branch" ];
  };
```

Reply:
Ah I see! Okay getting closer. I set the version to yesterday and tried to see what would happen when I ran the following command: `nix-shell maintainers/scripts/update.nix --argstr package reindeer`. Seems like I need to tell it that I need the `Cargo.lock` file to be generated because it does not exist.

```
Running update for:
 - reindeer-unstable-2023-05-17: UPDATING ...
 - reindeer-unstable-2023-05-17: ERROR

--- SHOWING ERROR LOG FOR reindeer-unstable-2023-05-17 ----------------------

this derivation will be built:
  /nix/store/b3hm4sfh9pfn6gbxxsq1qyk7jiwnb164-reindeer-unstable-2023-05-18.drv
building '/nix/store/b3hm4sfh9pfn6gbxxsq1qyk7jiwnb164-reindeer-unstable-2023-05-18.drv'...
error: builder for '/nix/store/b3hm4sfh9pfn6gbxxsq1qyk7jiwnb164-reindeer-unstable-2023-05-18.drv' failed with exit code
 1;
       last 5 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/qhiyv03qifwzdhg3z0yhvmzbklz7abfx-source
       > source root is source
       > cp: cannot stat 'source/Cargo.lock': No such file or directory
       > /nix/store/crv0aygayvds8xwyqln2p1zrdsmfw54s-stdenv-darwin/setup: line 145: pop_var_context: head of shell_vari
ables not a function context
```

The [nix-update](https://github.com/Mic92/nix-update) repository has some interesting notes, but nothing related to specific "generate a `Cargo.lock` if it does not exist" behavior.

I also see in the [Rust documentation for nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/rust.section.md) that `allowBuiltinFetchGit` is "for usage outside nixpkgs". I presume that won't help us solve the issue here and that it is unrelated.

Reply:
This seems to be the function that it hits... investigating.
https://github.com/Mic92/nix-update/blob/e242e61cec80802334b6d57530196d79f68bb6f1/nix_update/update.py#L159-L178

Reply:
ah I just realized, nix-update only supports it when the repository contains a `Cargo.lock` file. Sorry for the confusion, I think you would have to have a separate update script for now. Or just not have one and manually update it, that would work too

Reply:
Ah all good! I've filed an issue: https://github.com/Mic92/nix-update/issues/164. Might even take a crack at it, if I find the time.

Alright I'll add an update script and get back to you. Thanks for the help!

Reply:
Thanks!